packages/material-ui/src/ButtonBase/ButtonBase.js

```diff
@@ -99,82 +121,115 @@ class ButtonBase extends React.Component {
         ].join(' '),
       );
     }
-    prepareFocusVisible(button.ownerDocument);
+    return button.ownerDocument;
+  }, []);
+  const { isFocusVisible, onBlurVisible } = useIsFocusVisible(getOwnerDocument);
 
-    if (this.props.action) {
-      this.props.action({
+  React.useEffect(() => {
+    if (action) {
+      action({
         focusVisible: () => {
-          this.setState({ focusVisible: true });
-          this.getButtonNode().focus();
+          setFocusVisible(true);
+          buttonRef.current.focus();
         },
       });
     }
-  }
+  }, [action, buttonRef]);
 
-  componentDidUpdate(prevProps, prevState) {
-    if (
-      this.props.focusRipple &&
-      !this.props.disableRipple &&
-      !prevState.focusVisible &&
-      this.state.focusVisible
-    ) {
-      this.ripple.pulsate();
+  React.useEffect(() => {
+    if (focusVisible && focusRipple && !disableRipple) {
+      rippleRef.current.pulsate();
     }
-  }
+  }, [disableRipple, focusRipple, focusVisible]);
 
-  getButtonNode() {
-    return ReactDOM.findDOMNode(this.buttonRef.current);
-  }
+  function useRippleHandler(rippleAction, eventCallback, skipRippleAction = disableTouchRipple) {
+    function handleClient(event) {
+      if (eventCallback) {
+        eventCallback(event);
+      }
 
-  onRippleRef = node => {
-    this.ripple = node;
-  };
+      const { current: ripple } = rippleRef;
+      const ignore = event.defaultPrevented || skipRippleAction;
+      if (!ignore && ripple) {
+        ripple[rippleAction](event);
+      }
 
-  static getDerivedStateFromProps(nextProps, prevState) {
-    if (typeof prevState.focusVisible === 'undefined') {
-      return {
-        focusVisible: false,
-        lastDisabled: nextProps.disabled,
-      };
+      return true;
     }
 
-    // The blur won't fire when the disabled state is set on a focused input.
-    // We need to book keep the focused state manually.
-    if (!prevState.prevState && nextProps.disabled && prevState.focusVisible) {
-      return {
-        focusVisible: false,
-        lastDisabled: nextProps.disabled,
-      };
+    function handleServer() {
+      return () => {};
     }
 
-    return {
-      lastDisabled: nextProps.disabled,
-    };
+    return useEventCallback(typeof window === 'undefined' ? handleServer : handleClient);
```

Comment:
`typeof window === 'undefined'` misses the point of the original motivation:
https://github.com/mui-org/material-ui/blob/1291dcaa84e53265de166fcf481128ce3df25821/packages/material-ui/src/ButtonBase/createRippleHandler.js#L30-L32

The equivalent would be:

```suggestion
    if (typeof window === 'undefined') {
      return handleServer
    }

    // eslint-disable-next-line react-hooks/rules-of-hooks
    return useEventCallback(handleClient);
```

If you don't like that, let's just remove the "perf" branch logic. I don't even know if it's meaningful with this rewrite.

Reply:
The current code is equivalent to the old one. You just remove a useEventCallback call (virtually zero cost) and introduce (technically) conditional hooks call. Not sure what is gained by your suggestion.

Reply:
Ah no we replaced the whole `createRippleHandle` not just the return value.

```js
function handleServer() {}
``` 

it is then.

Reply:
Come to think of it I don't think that ever caused any meaningful gains. The events aren't dispatched on the server anyway so handlServer isn't called anyway.